bucket/drivers/s3/s3.go

```diff
@@ -120,19 +136,143 @@ func (d *S3Driver) DeleteBucket(id string, region string, clearBucket bool) erro
+		"AccessKeyId":     accessKeyResult.AccessKey.AccessKeyId,
+		"SecretAccessKey": accessKeyResult.AccessKey.SecretAccessKey,
+	}
+	accessKeyMarshalled, err := json.Marshal(accessKeyMap)
+	if err != nil {
+		return "", fmt.Errorf("unable to decode the credentials for account %s: %v", accountName, err)
+	}
+	credentials := string(accessKeyMarshalled[:])
+	return credentials, nil
+}
+
+// getUserPolicyInput gets the policy input that can be applied to the iam account
+func getUserPolicyInput(bucketId string, accountName string, inputAccessPolicy string, effect string) (*iam.PutUserPolicyInput, error) {
+	accessPolicy := inputAccessPolicy
+	// If acces policy is not specified, use default access policy for the bucket
+	if len(accessPolicy) == 0 {
+		bucketPolicy := map[string]interface{}{
+			"Statement": []map[string]interface{}{
+				{
+					"Effect": effect,
+					"Action": []string{
+						"s3:*",
+					},
+					"Resource": []string{
+						fmt.Sprintf("arn:aws:s3:::%s/*", bucketId),
+						fmt.Sprintf("arn:aws:s3:::%s", bucketId),
+					},
+				},
+			},
+		}
+		policy, err := json.Marshal(bucketPolicy)
+		if err != nil {
+			return nil, err
+		}
+		accessPolicy = string(policy[:])
+	}
+	input := &iam.PutUserPolicyInput{
+		// PolicyDocument: aws.String("{\"Statement\":{\"Effect\":\"Allow\",\"Action\":\"*\",\"Resource\":\"*\"}}"),
+		PolicyDocument: aws.String(accessPolicy),
+		PolicyName:     aws.String(bucketId + "-AccessPolicy"),
+		UserName:       aws.String(accountName),
+	}
+	return input, nil
+}
+
+// grantAccessToBucket allows iam account access to the bucket
+func grantAccessToBucket(bucketId string, accountName string, accessPolicy string, iamSvc *iam.IAM) error {
+	input, err := getUserPolicyInput(bucketId, accountName, accessPolicy, "Allow")
+	if err != nil {
+		return err
+	}
+	_, err = iamSvc.PutUserPolicy(input)
+	if err != nil {
+		return fmt.Errorf("unable to provide user %s access to bucket %s : %v", accountName, bucketId, err)
+	}
+	return nil
+}
+
+// revokeAccessToBucket revoks iam account access to the bucket
+func revokeAccessToBucket(bucketId string, accountName string, iamSvc *iam.IAM) error {
+	input, err := getUserPolicyInput(bucketId, accountName, "", "Deny")
+	if err != nil {
+		return err
+	}
+	_, err = iamSvc.PutUserPolicy(input)
+	if err != nil {
+		return fmt.Errorf("unable to revoke user %s access to bucket %s : %v", accountName, bucketId, err)
+	}
+	return nil
+}
+
 // AccessBucket grants access to the S3 bucket
-// Dummy impplementation
-// Actual implementation to be done once we have more clarity on the downstream API
 func (d *S3Driver) GrantBucketAccess(id string, accountName string, accessPolicy string) (string, string, error) {
-	logrus.Info("bucket_driver.S3 access bucket received")
-	return accountName, "", nil
+	iamSvc, err := d.NewIamSvc()
+	if err != nil {
+		return "", "", fmt.Errorf("unable to create iam session: %v ", err)
+	}
+
+	accountId, err := createUser(accountName, iamSvc)
```

Comment:
Currently we are not deleting the user. We just deny access to the resource when we get BucketRevoke request. We can discuss about how we want to deal with it and it can be implemented as an incremental improvement.